internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala

```diff
@@ -64,7 +66,38 @@ object Locate {
     className => if (defClass(className)) getF(className).toRight(true) else Left(false)
   }
 
-  def definesClass(entry0: VirtualFile): DefinesClass =
+  /**
+   * Can be used to determine whether or not a given file has a class definition
+   * with a particular name. It will cache the set of class names for each jar file.
+   * Each jar cache entry is invalidated when the last modified time of the jar file
+   * changes.
+   *
+   */
+  sealed abstract class DefinesClass {
+    def apply(entry0: VirtualFile): XDefinesClass
+  }
+  class CachingDefinesClass extends DefinesClass {
+    private[this] val jarCache = new ConcurrentHashMap[Path, (JarDefinesClass, Long)]
```

Comment:
This makes me a bit nervous because there could be thousands of JAR files. Would it make sense to use Caffeine (or equivalent) that could put upper limit on the cache size and keep everything in a weak reference?

Reply:
Possibly. My guess though is that if the classpath is thousands of files that this cache will still represent a small fraction of the total memory of the process running zinc.

Reply:
Unless we shaded caffeine, we probably shouldn't put it in Zinc so we'd need to do it at the sbt level. It's a bit tricky to rework the api because I'd probably need to expose `JarDefinesClass`.

When I tested this in sbt, I added a setting `cacheClasspathEntries` that was enabled by default but could be disabled. Having the binary switch seems better for now than trying to tune the size of a caffeine cache.

That being said, it makes me wonder if we should have something that's the opposite of turbo mode that disables as many of the caches as possible.

Reply:
Also, it's worth noting that the default behavior in zinc is not cached so this shouldn't impact any downstream zinc users unless they opt in.

Reply:
Looking at sbt side I see:

```scala
      classpathEntryDefinesClass := {
        val converter = fileConverter.value
        val f = FileValueCache({ x: NioPath =>
          if (x.getFileName.toString != "rt.jar") Locate.definesClass(converter.toVirtualFile(x))
          else ((_: String) => false): DefinesClass
        }).get;
        { (x: File) =>
          f(x.toPath)
        }
      },
```

I guess the facility of caching per file is already there, but it's limited to per session because it's a task?